bottomless2/src/lib.rs

```diff
@@ -0,0 +1,179 @@
+#![allow(dead_code, unused_variables, async_fn_in_trait)]
+use std::path::PathBuf;
+use std::time::Duration;
+
+use chrono::{DateTime, Utc};
+use tokio::sync::{mpsc, oneshot};
+use tokio::task::{JoinHandle, JoinSet};
+
+use self::job::JobResult;
+use self::scheduler::Scheduler;
+use self::storage::Storage;
+
+mod job;
+mod restore;
+mod scheduler;
+pub mod storage;
+
+/// Backgroung loop task state.
+///
+/// The background loop task is not allowed to exit, unless it was notified for shutdown.
+///
+/// On shutdown, attempts to empty the queue, and flush the receiver. When the last handle of the
+/// receiver is dropped, and the queue is empty, exit.
+pub struct BottomlessLoop<S: Storage> {
+    receiver: mpsc::Receiver<StoreSegmentRequest<S::Config>>,
+    scheduler: Scheduler<S>,
+    max_in_flight: usize,
+    in_flight_futs: JoinSet<JobResult<S>>,
+    force_shutdown: oneshot::Receiver<()>,
+}
+
+impl<S: Storage + 'static> BottomlessLoop<S> {
+    /// Schedules durability jobs. This loop is not allowed to fail, or loose jobs.
+    /// A job is prepared by calling `Scheduler::prepare(..)`. The job is spawned, and it returns a
+    /// `JobResult`, which is then returned to the scheduler by calling `Scheduuler::report(..)`.
+    /// When a request is received, it is immediately scheduled by calling `Scheduler::register`
+    /// with it.
+    ///
+    /// The loop is only allowed to shutdown if the receiver is closed, and the scheduler is empty,
+    /// or if `force_shutdown` is called, in which case everything is dropped in place.
+    #[tracing::instrument(skip(self))]
+    async fn run(mut self) {
+        let mut shutting_down = false;
+        // run the loop until shutdown.
+        loop {
+            if shutting_down && self.scheduler.is_empty() {
+                break;
+            }
+
+            // schedule as much work as possible
+            while self.scheduler.has_work() && self.in_flight_futs.len() < self.max_in_flight {
+                let job = self
+                    .scheduler
+                    .schedule()
+                    .expect("scheduler has work, but didn't return a job");
+                self.in_flight_futs.spawn(job.perform());
+            }
+
+            tokio::select! {
+                msg = self.receiver.recv(), if !shutting_down => {
+                    match msg {
+                        Some(req) => {
+                            self.scheduler.register(req);
+                        }
+                        None => {
+                            shutting_down = true;
+                        }
+                    }
+                }
+                Some(join_result) = self.in_flight_futs.join_next(), if !self.in_flight_futs.is_empty() => {
```

Comment:
I think we should consider making this a `biased` (info in https://docs.rs/tokio/latest/tokio/macro.select.html) so that these always get processed first instead of enqueuing things to prevent starvation. If the mpsc queue is always full it can be trying to enqueue more where I think we should always prioritize getting the results out first.